compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -2104,6 +2104,98 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
                         );
                     }
                 }
+
+                if let (Some(body_id), Some(ty::subst::GenericArgKind::Type(_))) =
+                    (body_id, subst.map(|subst| subst.unpack()))
+                {
+                    struct FindExprBySpan<'hir> {
```

Comment:
I wonder if we could introduce a new `ObligationCauseCode` that stores the hir id of the `ExprKind::Path` that this obligation comes from, so we wouldn't need to use this span-based search method.

Reply:
That sounds like a good idea.